tests/phpunit/CRM/Contact/BAO/GroupContactCacheTest.php

```diff
@@ -188,6 +159,100 @@ public function assertCacheMatches($expectedContactIds, $groupId) {
     $this->assertEquals($expectedContactIds, $actualContactIds);
   }
 
+  /**
+   * Test the opportunistic refresh cache function does not touch non-expired entries.
+   */
+  public function testOpportunisticRefreshCacheNoChangeIfNotExpired() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    $this->assertCacheMatches(
+      array($deceased[0]->id, $deceased[1]->id, $deceased[2]->id),
+      $group->id
+    );
+    CRM_Contact_BAO_GroupContactCache::opportunisticCacheFlush();
+
+    $this->assertCacheNotRefreshed($deceased, $group);
+  }
+
+  /**
+   * Test the opportunistic refresh cache function does refresh stale entries.
+   */
+  public function testOpportunisticRefreshChangeIfCacheDateFieldStale() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    CRM_Core_DAO::executeQuery('UPDATE civicrm_group SET cache_date = DATE_SUB(NOW(), INTERVAL 7 MINUTE) WHERE id = ' . $group->id);
```

Comment:
A general thought for future reference: when testing time-related code, it's useful if all the timestamps are generated via `CRM_Utils_Time` instead of `NOW()` or `time()`. That allows you to write tests which manipulate the clock - and there's no need to poke inside the cache data-structure. (The only limitation - you can't use `CRM_Utils_Time` within a SQL trigger, but I don't think we use that much outside of the logging-tables.)